### DISCUSSION TITLE
Importance of Rank, Alpha, LR, and Optimizer

### DISCUSSION THREAD

**Discussion by TheInfinites** [bmaltais/kohya_ss#3131] [General]
I want to train a character lora. I really want the character to be super consistent, just like the dataset. 
I'm using someone else's config file. They have Rank and Alpha set to 128. LR to 0.00005. And the optimizer is Adafactor. 
I did some research myself (mainly using Chatgpt) and, to be honest, I'm ultra confused. 
I just want to know how much importance these settings play in the training. If someone here knows about these things, it would be great.

**Comment by 5KilosOfCheese** [comment]
Every training is unique, and every base model trains differently. Every dataset and type of concept requires it's own adjustments in the parameters. Even the batch size effects the results dramatically (By my experience on Flux bigger batch size tamed the learning process, while on SDXL it felt as if it made it more aggressive). There are no universal settings, other than those that produce a corret output in the sense that it functions; once you have that it is more or less finding the correct settings to adjust things.

Flux for example trains VERY aggressively and quickly, compared to something like SDXL. AdamW8bit seems to go well with polynomial LR curve. Adafactor with constant. And with Flux you'll need different LR based on your Lora dimensions (rank). I personally don't use any regulation because I keep getting best results without it.

I am fairly good at intuitively going about figuring out how the settings should be adjusted, however there are still many which are total unknowns for me, and therefor I don't touch them.

However don't consult ChatGPT. Just read the documentation and published papers, they explain what is going on. Another good source is the pull requests on the Kohya's repo which explain the code that is being pushed, check the discussion. These are the actual resources I have used. The worst resource you can use are the clickbait youtube videos, keyword optimised SEO blog posts, and anyone who locks things behind paywalls. If you actually want to get good at training stuff, you'll need to experiment, you'll need to analyse and you'll need to read documentation. Nothing that happens under the hood is a secret or a mystery, granted it can be arcane and obscure at time, but everything from lowest to highest level is something you can figure out if you want to. 

Also here is a fact: High dimension count (Rank) is not automatically better; you'll get better results at lower ranks that are properly trained.

The approach I use is that I start at fairly low dimensions, like... 4 or 8; with Flux (Which is what I been using mostly lately) it is very obvious if you hit the "information limit" of ranks. Once I have something which clearly has the broadest concept of the thing I am trying to train, I add more dimensions, and see where that takes me. I generally add half the current rank, so from 8 I go to 12, 12 to 18... etc. This is just how I do it.

But here is the basics: 
Dimensions (Rank) = the amount of information the LoRA can store, this also quite literally increases the file size and memory demand for training and use.

Alpha = the multiplier of the information stored within the LoRA. How this behave has thus far been proven to be extremely base model architecture specific. SD1, SD2, SDXL, and Flux (I have not tried SD3.5) have all had different behavior when it comes to the Alpha. However the rule of thumb is that if you do a test training, and getting good results means that you need to lower the LoRA power in the generation process, you probably could benefit from lower Alpha during training; and vice versa if you need to increase the LoRA power. Since the LoRA power significantly alters the generation process, it more beneficial (in my opinion) to get it close to correct with Alpha during training.

LR = Is just how much information is altered during the training process. Too low of an LR means the training will never converge to a solution. You can imagine the training being a journey through valleys, we want to find the correct place in the valleys, too low LR means you ain't going anywhere, too high LR means you'll miss your mark and keep going further away from it, even possibly you can go outside of the model's "universe", which will lead to the LoRA generating absolute nonsense.

Optimiser: This is just the method of math being utilised to calculate things during the training process. AdamW is the classic and generally "the thing which will work", most of the ones available in the GUI are just different versions of Adam or Adam + something. Adafactor for example is Adam + _Adaptive Learning Rate_ in a specific memory efficient way.  Which is the best? People will fight over this endlessly, and defend their own favourite; however not all work on all base models, but some work basically for everything; they all do the same thing, in a different way. Some need special settings and adjustments, some are very memory intensive, some are not; others are really slow and converge with few steps, some are quick and need more steps.

**Reply by TheInfinites** [reply]
Thanks a lot [5KilosOfCheese](https://github.com/5KilosOfCheese) for explaining in such detail. I really appreciate it. 

I just have a few questions. (and I'm using Flux as my base model)
Firstly, you mentioned that
> However the rule of thumb is that if you do a test training, and getting good results means that you need to lower the LoRA power in the generation process

Do you mind clarifying this a bit more? I usually keep my Rank and Alpha the same (128) I understand that this highly depends on the number of images you have. 

Secondly, how can one go about deciding the right LR?
In my recent training (rank and alpha were 64), my LR was 0.00005. My Optimizer was Adafactor. From what I've heard, Adafactor automatically changes or rather updates the LR. But I'm planning on testing different ones. Next on the list is AdamW and Prodigy. 

Lastly, when it comes to Epochs, what would you say is the rule of thumb? is 3k steps fine for Character lora?

Basically, the next training I will be doing has 30 images in the dataset. I'll be going with rank/alpha 128. Around 6k steps. Optimizer AdamW. The LR is something I'm a bit confused about. 

Edit: Added the last question.

**Reply by 5KilosOfCheese** [reply]
>Do you mind clarifying this a bit more? I usually keep my Rank and Alpha the same (128) I understand that this highly depends on the number of images you have.

No. These have to do with the properties of the model's contents. Higher ranks means more information. Alpha however is bit dependant on what the model is for, and what base model it is for. Alpha is just the multiplier of strenght. This whole "Ranks are tied to amoung of images" is just one of those old wives tales. It's better to think it as "More there is diversity of the concept in the dataset, the higher the rank of the Lora should be" if you have a simple concept amongs otherwise complex images, you'll just end up packing in information that is unnecessary and at times reduce the flexibility and quality of the output.  The training process will try to pack in as much generalised average it can get from the dataset. You can actually train it to basically complete replicate an image if you want to - within a specific margin - but this isn't what you really are trying to do. Right? Estimating how big the rank needs to be is absurdly difficult, but not impossible. Keep in mind that during the generation process the missing information will be put in from the base model. So if you train fabric object, and the texture doesn't get included, the generation will try to put it in (because that is what it does inherently) from what it already knows. This will lead to things looking weird. Like I had a problem where I tried to train things made with gauze, and the generation ended up making it a lace pattern, fixing this took so backwards moon logic, but it was doable.

The right LR is quite simple, if you want to make it. If at the end the last resulting model is _weak_  then the LR is too low (however the LR-curve affects things a bit). If the end result is too _aggressive_ then the LR was too high. However with Flux there is yet another weird link - apparently - is that lower your ranks, higher your LR should be; and as you increase ranks you should lower the LR. Which does make sense in a way. The problem is that if your LR is too low, the training system will basically disregard the training all together and doesn't write anything really into the model. If your LR is too high it write aggressive changes than can push the generation process out of bounds for the model, and you just get random colour or black.

For question of epoch. I usually hang around 10-30. If it is clear it isn't working at 20, then it is not working at all. I generally train 10 epochs at a time. This usually is equivalent of 1000-3000 steps, the way I like to go about thing.

As for character LoRAs specifically? I don't actually do any of those so I can't comment. I also don't do any like "anime"/cartoon/etc. stuff. There is a HUGE difference between realisms and not-realism when it comes to training a thing, I have trained few cartoon objects to get them to be "real objects" and I found this incredibly difficult, however the reverse (Real to cartoon) was piss easy. I have managed to re-create faces present in the dataset with very few steps and low ranks, and sometimes I can't even get a close approximation of the subject to appear at all. 

My approach is that of an careful engineer - because I am one in real life. I fuck around a lot with powerful lasers, that used in welding and fabrication. The amount of parameters I have available, and which all affect the output quality is about twice as what we have available in this GUI/LoRA training - and some of those can actually fuck things up for real. So I never work on more than 3-4 parameters at once, and depending on the type of parameter, I take either ±50% steps or ±10% steps, and when I hit a limit of "this doesn't work anymore" I go backwards the delta of last change. I have found that is actually extremely good approach to use in something like LoRAs. I select few parameters. I have actually tried LoRAs with ranks low as 2, out of curiosity really. I have figured out that the sweet spot for me is between 8-18, rarely do I need to venture outside of those. Below 8 tends to be best for general broad "feels" of an image; above 18 I tend to get useless shit included.  **Because we are talking about flux and as established earlier ranks affect optimal LR.** I generaly keep 0,001 for rank of 16, and scale from there.

**HOWEVER** choosing LR is also dependant on the optimiser and scheduler you use. Adaptive optimisers behave differently as they adjust the LR constnatly (generally downwards). Prodigy (which I like fair bit) needs to have LR = 1, and should only be changed if there is instability (This even reads on the console when you start training with prodigy below LR = 1). On Flux AdamW8Bit + Polynomial power 2, starting from 0.004 in my experience is a really good for capturing the general shape and structure of things, and some of the colour palette (with 20 epochs, ~20 images). Make of that what you want.

**Reply by HuskedHull** [reply]
> For question of epoch. I usually hang around 10-30. If it is clear it isn't working at 20, then it is not working at all. I generally train 10 epochs at a time. This usually is equivalent of 1000-3000 steps, the way I like to go about thing.

Repeats, bucket* size?

*I mean batch size

**Reply by 5KilosOfCheese** [reply]
> > For question of epoch. I usually hang around 10-30. If it is clear it isn't working at 20, then it is not working at all. I generally train 10 epochs at a time. This usually is equivalent of 1000-3000 steps, the way I like to go about thing.
> 
> Repeats, bucket size?

My repeats vary from 5-10 depending on many factors - rarely I go beyond that. I scale all images (on flux) to 512x512, or break down bigger images into those segments. I have not experienced any real benefit or downside from other resolutions; so... I don't bother. Since I use only one resolution I don't bucket anything. 

I generally edit the images so that I can fit them into that resolution; however if it is truly impossible. Bucket steps are 64px - the default in this repos tested setting; min 256, max 1024. 

Batch size is 1-4, once again depending on how much spare VRAM I got left since different settings take different amounts of VRAM.

Generally I train everything in a square format. I had problems with weird borders being generated, and instead of trying to solve this, I just decided to not bother. When I wanted to make sure details rally made it in, I just trained square images from 1024 to 512, going down 64 steps. So batch size 1.

Every training I do, has basically been unique in it's configuration. I got a baseline settings from which I start, but I always endup tweaking things. My dataset size have ranged from 5-50.

Seriously... Just find a setting that gets some results, and start experimenting and tweaking.
